dot/state/block.go

```diff
@@ -892,14 +896,79 @@ func (bs *BlockState) HandleRuntimeChanges(newState *rtstorage.TrieState,
 		return fmt.Errorf("failed to update code substituted block hash: %w", err)
 	}
 
-	newVersion := rt.Version()
+	newVersion := parentRuntimeInstance.Version()
 	go bs.notifyRuntimeUpdated(newVersion)
 	return nil
 }
 
 // GetRuntime gets the runtime instance pointer for the block hash given.
 func (bs *BlockState) GetRuntime(blockHash common.Hash) (instance Runtime, err error) {
-	return bs.bt.GetBlockRuntime(blockHash)
+	// we search primarily in the blocktree so we ensure the
+	// fork aware property while searching for a runtime, however
+	// if there is no runtimes in that fork than we look for the
+	// closest ancestor with a runtime instance
+	runtimeInstance, err := bs.bt.GetBlockRuntime(blockHash)
+	if errors.Is(err, blocktree.ErrRuntimeNotFound) {
+		return bs.closestAncestorWithInstance(blockHash)
+	}
+
+	// in this case the node is not in the blocktree which mean
+	// it is a finalized node already persisted in database, so we
+	// should check if it is in the mapping or create a instance for it
+	if errors.Is(err, blocktree.ErrNodeNotFound) {
+		// TODO: https://github.com/ChainSafe/gossamer/issues/3066
+		panic("TODO: while getting a runtime but node does not exists")
+	}
+
+	return runtimeInstance, err
+}
+
+func (bs *BlockState) closestAncestorWithInstance(blockHash common.Hash) (instance Runtime, err error) {
+	allHashesInMapping := bs.bt.GetInMemoryRuntimesBlockHashes()
+	if len(allHashesInMapping) == 0 {
+		panic("no runtime instances available")
+	}
+
+	if len(allHashesInMapping) == 1 {
+		return bs.bt.GetBlockRuntime(allHashesInMapping[0])
+	}
+
+	allHeaders := make([]types.Header, len(allHashesInMapping))
+	for idx, hashInMapping := range allHashesInMapping {
+		header, err := bs.GetHeader(hashInMapping)
+		if err != nil {
+			return nil, fmt.Errorf("getting header: %w", err)
+		}
+
+		allHeaders[idx] = *header
+	}
+
+	if len(allHeaders) > 1 {
+		// sort the slice using the descending order
+		sort.Slice(allHeaders, func(i, j int) bool {
+			return allHeaders[i].Number > allHeaders[j].Number
+		})
+	}
```

Comment:
You should cache this, or just keep the sorted list in `BlockTree` if you're going to be doing this every time `GetRuntime()` is called.